File: 2022/day05/clojure/solution.clj
#!/usr/bin/env bb

(require '[clojure.string :as str])

(defn parse-stacks [stack-section]
  "Parse the visual crate diagram into vectors of crates (bottom to top)"
  (let [lines (str/split-lines stack-section)
        ;; Last line has stack numbers - count them
        num-stacks (count (str/split (str/trim (last lines)) #"\s+"))
        ;; Process all lines except the last (numbers line)
        crate-lines (butlast lines)]
    ;; For each stack position, collect crates from top to bottom, then reverse
    (->> (range num-stacks)
         (mapv (fn [stack-idx]
                 (let [pos (+ 1 (* stack-idx 4))] ; Position of crate letter
                   (->> crate-lines
                        (keep (fn [line]
                                (when (< pos (count line))
                                  (let [c (nth line pos)]
                                    (when (not= c \space)
                                      c)))))
                        reverse
                        vec)))))))

(defn parse-moves [move-section]
  "Parse move instructions into [count from-stack to-stack] tuples (0-indexed)"
  (->> (str/split-lines move-section)
       (keep (fn [line]
               (when-let [match (re-find #"move (\d+) from (\d+) to (\d+)" line)]
                 [(parse-long (nth match 1))
                  (dec (parse-long (nth match 2)))  ; 0-indexed
                  (dec (parse-long (nth match 3)))])))))

(defn parse-input [filename]
  (let [content (slurp filename)
        [stack-section move-section] (str/split content #"\n\n")]
    [(parse-stacks stack-section)
     (parse-moves move-section)]))

(defn part1 [stacks moves]
  "Move crates one at a time (reverses order)"
  (let [result-stacks
        (reduce (fn [stacks [n from-stack to-stack]]
                  ;; Move 'n' crates one at a time
                  (loop [stacks stacks
                         remaining n]
                    (if (zero? remaining)
                      stacks
                      (let [crate (peek (nth stacks from-stack))]
                        (recur (-> stacks
                                   (update from-stack pop)
                                   (update to-stack conj crate))
                               (dec remaining))))))
                stacks
                moves)]
    (apply str (map peek result-stacks))))

(defn part2 [stacks moves]
  "Move multiple crates at once (preserves order)"
  (let [result-stacks
        (reduce (fn [stacks [n from-stack to-stack]]
                  (let [from-vec (nth stacks from-stack)
                        split-point (- (count from-vec) n)
                        crates-to-move (subvec from-vec split-point)
                        remaining (subvec from-vec 0 split-point)]
                    (-> stacks
                        (assoc from-stack remaining)
                        (update to-stack into crates-to-move))))
                stacks
                moves)]
    (apply str (map peek result-stacks))))

(defn -main []
  (let [script-dir (-> *file* java.io.File. .getParent)
        input-file (str script-dir "/../input.txt")
        [stacks moves] (parse-input input-file)]
    (println "Part 1:" (part1 stacks moves))
    (println "Part 2:" (part2 stacks moves))))

(-main)

File: 2022/day05/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class Solution {

    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get(System.getProperty("user.dir"), "..", "input.txt");
        String content = Files.readString(inputPath);

        String[] parts = content.split("\n\n");
        String[] stackLines = parts[0].split("\n");
        String[] moveLines = parts[1].trim().split("\n");

        // Find number of stacks from the last line (the numbers)
        int numStacks = stackLines[stackLines.length - 1].trim().split("\\s+").length;

        // Parse stacks (bottom-up, excluding the number line)
        List<List<Character>> stacks = new ArrayList<>();
        for (int i = 0; i < numStacks; i++) {
            stacks.add(new ArrayList<>());
        }

        for (int lineIdx = 0; lineIdx < stackLines.length - 1; lineIdx++) {
            String line = stackLines[lineIdx];
            for (int i = 0; i < numStacks; i++) {
                int pos = 1 + i * 4;  // Position of crate letter
                if (pos < line.length() && line.charAt(pos) != ' ') {
                    stacks.get(i).add(line.charAt(pos));
                }
            }
        }

        // Reverse so bottom is at index 0
        for (List<Character> stack : stacks) {
            Collections.reverse(stack);
        }

        // Parse moves
        List<int[]> moves = new ArrayList<>();
        Pattern pattern = Pattern.compile("move (\\d+) from (\\d+) to (\\d+)");
        for (String line : moveLines) {
            Matcher matcher = pattern.matcher(line);
            if (matcher.find()) {
                int count = Integer.parseInt(matcher.group(1));
                int fromStack = Integer.parseInt(matcher.group(2)) - 1;  // 0-indexed
                int toStack = Integer.parseInt(matcher.group(3)) - 1;
                moves.add(new int[]{count, fromStack, toStack});
            }
        }

        System.out.println("Part 1: " + part1(stacks, moves));
        System.out.println("Part 2: " + part2(stacks, moves));
    }

    private static List<List<Character>> deepCopy(List<List<Character>> stacks) {
        List<List<Character>> copy = new ArrayList<>();
        for (List<Character> stack : stacks) {
            copy.add(new ArrayList<>(stack));
        }
        return copy;
    }

    private static String part1(List<List<Character>> originalStacks, List<int[]> moves) {
        List<List<Character>> stacks = deepCopy(originalStacks);

        for (int[] move : moves) {
            int count = move[0];
            int fromStack = move[1];
            int toStack = move[2];

            for (int i = 0; i < count; i++) {
                List<Character> from = stacks.get(fromStack);
                char crate = from.remove(from.size() - 1);
                stacks.get(toStack).add(crate);
            }
        }

        StringBuilder result = new StringBuilder();
        for (List<Character> stack : stacks) {
            if (!stack.isEmpty()) {
                result.append(stack.get(stack.size() - 1));
            }
        }
        return result.toString();
    }

    private static String part2(List<List<Character>> originalStacks, List<int[]> moves) {
        List<List<Character>> stacks = deepCopy(originalStacks);

        for (int[] move : moves) {
            int count = move[0];
            int fromStack = move[1];
            int toStack = move[2];

            List<Character> from = stacks.get(fromStack);
            List<Character> to = stacks.get(toStack);

            // Move multiple crates at once (preserve order)
            List<Character> crates = new ArrayList<>(from.subList(from.size() - count, from.size()));
            for (int i = 0; i < count; i++) {
                from.remove(from.size() - 1);
            }
            to.addAll(crates);
        }

        StringBuilder result = new StringBuilder();
        for (List<Character> stack : stacks) {
            if (!stack.isEmpty()) {
                result.append(stack.get(stack.size() - 1));
            }
        }
        return result.toString();
    }
}
